internal/compiler-interface/src/main/java/xsbti/VirtualFileRef.java

```diff
@@ -0,0 +1,113 @@
+/*
+ * Zinc - The incremental compiler for Scala.
+ * Copyright Lightbend, Inc. and Mark Harrah
+ *
+ * Licensed under Apache License 2.0
+ * (http://www.apache.org/licenses/LICENSE-2.0).
+ *
+ * See the NOTICE file distributed with this work for
+ * additional information regarding copyright ownership.
+ */
+
+package xsbti;
+
+/**
+ * <code>VirtualFileRef</code> represents a reference to a file-like object.
+ * Unlike <code>java.io.File</code> or <code>java.nio.file.Path</code>
+ * that are tied to filesystem-related capabilities, <code>VirtualFileRef</code>
+ * is designed to be a less-capable pointer to an object whose
+ * main feature is having a path-like identity.
+ * The equality is expected to be implemented using the equality
+ * of the <code>id()</code> alone. See {@link BasicVirtualFileRef}.
+ *
+ * <p>{@link VirtualFile}, a subtype of <code>VirtualFileRef</code>
+ * supports minimally viable file-like opration needed for compilation.
+ * To illustrate the difference between the two, consider the
+ * following flow of operation.
+ * </p>
+ *
+ * <h3>Flow of operation</h3>
+ * <p>Suppose that there are two people Alice and Bob who are working on the same repo.
+ * On Alice's machine, the build tool would construct
+ * an instance of {@link FileConverter} used for the entire build.
+ * The reference implementation is <code>sbt.internal.inc.MappedFileConverter</code>.
+ * The build tool would pass the usual suspect of absolute paths
+ * to this <code>MappedFileConverter</code> such as the base directory
+ * of the working directory, Coursier cache directory, etc.
+ * Given the sequence of Scala and Java source files,
+ * <code>MappedFileConverter</code> will create a sequence of
+ * <code>MappedVirtualFile</code> whose <code>id</code> would look like
+ * <code>${0}/src/main/example/A.scala</code>.
```

Comment:
I don't undestand what this `${0}` thing is. Why isn't the id `src/main/example/A.scala` ?

Reply:
`MappedFileConverter` can calculate the id based on multiple root paths, including for example Coursier cache directory.

Reply:
if toPath on a MappedVirtualFile does not resolve the id what should you substitute?